compiler-core/src/type_/expression.rs

```diff
@@ -202,12 +202,30 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
         }
     }
 
-    fn infer_panic(&mut self, location: SrcSpan, message: Option<EcoString>) -> TypedExpr {
+    fn infer_panic(
+        &mut self,
+        location: SrcSpan,
+        message: Option<Box<UntypedExpr>>,
+    ) -> Result<TypedExpr, Error> {
+        let message = message.map(|untyped_expr| self.infer(*untyped_expr));
         let typ = self.new_unbound_var();
-        TypedExpr::Panic {
-            location,
-            type_: typ,
-            message,
+        match message {
+            Some(Ok(message)) => {
+                // Making sure that the panic message evaluates to a string.
+                unify(string(), message.type_())
+                    .map_err(|e| convert_unify_error(e, message.location()))?;
+                Ok(TypedExpr::Panic {
+                    location,
+                    type_: typ,
+                    message: Some(Box::new(message)),
+                })
+            }
+            None => Ok(TypedExpr::Panic {
+                location,
+                type_: typ,
+                message: None,
+            }),
+            Some(Err(e)) => Err(e),
```

Comment:
I'm finding this a bit hard to read- could it be re-written without using `.map`? I think this would be clearer if it's used with just pattern matching as it means that `?` can be used directly and `Some(Err(e))` won't need to be handled.

```rust
let message = match message {
   ....
};
Ok(TypedExpr::Panic { ..... })
```

Reply:
Gotcha. Will try making this part easy to read.